crate/server/src/core/operations/revoke.rs

```diff
@@ -24,140 +25,173 @@ pub async fn revoke_operation(
+        .db
+        .retrieve(uid_or_tags, user, ObjectOperationType::Revoke, params)
+        .await?
+        .into_iter()
+        .filter(|owm| {
+            let object_type = owm.object.object_type();
+            (owm.state == StateEnumeration::Active || owm.state == StateEnumeration::PreActive)
+                && (object_type == ObjectType::PrivateKey
+                    || object_type == ObjectType::SymmetricKey
+                    || object_type == ObjectType::PublicKey)
+        })
+        .collect::<Vec<ObjectWithMetadata>>();
+
+    if owm_s.is_empty() {
+        return Err(KmsError::ItemNotFound(uid_or_tags.to_owned()))
+    }
+
+    // revoke the keys found
+    for owm in owm_s {
+        // perform the chain of revoke operations depending on the type of object
+        let object_type = owm.object.object_type();
+        match object_type {
+            SymmetricKey => {
+                // revoke the key
+                revoke_key_core(
+                    &owm.id,
+                    revocation_reason.clone(),
+                    compromise_occurrence_date,
                     kms,
-                    user,
                     params,
                 )
-                .await;
+                .await?;
             }
-            if let KeyFormatType::CoverCryptSecretKey = private_key.key_block()?.key_format_type {
-                revoke_user_decryption_keys(
-                    unique_identifier,
-                    request.revocation_reason,
-                    request.compromise_occurrence_date,
+            PrivateKey => {
+                //add this key to the ids to skip
+                ids_to_skip.insert(owm.id.clone());
+                // for Covercrypt, if that is a master secret key, revoke the user decryption keys
+                if let KeyFormatType::CoverCryptSecretKey = owm.object.key_block()?.key_format_type
+                {
+                    revoke_user_decryption_keys(
+                        &owm.id,
+                        revocation_reason.clone(),
+                        compromise_occurrence_date,
+                        kms,
+                        user,
+                        params,
+                        ids_to_skip.clone(),
+                    )
+                    .await?
+                }
+                // revoke any linked public key
+                if let Some(public_key_id) =
+                    owm.object.attributes()?.get_link(LinkType::PublicKeyLink)
+                {
+                    if !ids_to_skip.contains(&public_key_id) {
+                        recursively_revoke_key(
+                            &public_key_id,
+                            revocation_reason.clone(),
+                            compromise_occurrence_date,
+                            kms,
+                            user,
+                            params,
+                            ids_to_skip.clone(),
+                        )
+                        .await?;
```

Comment:
It seems to me that recursion could be avoided by checking for linked keys and then calling the revocation on each key found.

Also, it seems the revocation is performed on a per-ID basis. Does this imply that there is one network call per revoked key? Would it be possible to batch revocations?

If something like this can work, it would make the process simpler:
```
fn revoke_key(id):
    ids = get_linked_ids()
    revoke_key_core(ids)```

Reply:
I do not think you can avoid recursion to follow the key links and build the graph
Also, this is harder than it seems because there are loops in the graph: a private key links to a public key which links back to the private key

Reply:
The KMIP specification does not allow for batch (which is an issue, particularly on encrypt and decrypt)